src/module_5/module5.md

```diff
@@ -0,0 +1,5880 @@
+      <td>2020-03-31</td>
+    </tr>
+    <tr>
+      <th>39</th>
+      <td>0.618439</td>
+      <td>0.487939</td>
+      <td>39</td>
+      <td>2020-03-31</td>
+    </tr>
+  </tbody>
+</table>
+<p>2240 rows × 4 columns</p>
+</div>
+
+
+
+
+```python
+test_results_final_tree = test_results.sort_values(
+    ["execution_date", "n_trees"]
+).drop_duplicates("execution_date", keep="last")
+train_results_final_tree = train_results.sort_values(
+    ["execution_date", "n_trees"]
+).drop_duplicates("execution_date", keep="last")
+```
+
+And this are the results:
+
+
+```python
+(
+    ggplot(test_results_final_tree)
+    + geom_point(aes(x="execution_date", y="weighted-return"))
+    + theme(axis_text_x=element_text(angle=90, vjust=0.5, hjust=1))
+)
+```
+
+
+    
+![png](module5_files/module5_32_0.png)
+    
+
+
+
+
+
+    <Figure Size: (640 x 480)>
+
+
+
+
+```python
+(
+    ggplot(train_results_final_tree)
+    + geom_point(aes(x="execution_date", y="weighted-return"))
+    + theme(axis_text_x=element_text(angle=90, vjust=0.5, hjust=1))
+)
+```
+
+
+    
+![png](module5_files/module5_33_0.png)
+    
+
+
+
+
+
+    <Figure Size: (640 x 480)>
+
+
+
+We have trained the first models for all the periods for you, but there are a lot of things which may be wrong or can be improved. Some ideas where you can start:
+* Try to see if there is any kind of data leakage or suspicious features
+* If the training part is very slow, try to see how you can modify it to execute faster tests
+* Try to understand if the algorithm is learning correctly
+* We are using a very high level metric to evaluate the algorithm so you maybe need to use some more low level ones
+* Try to see if there is overfitting
+* Try to see if there is a lot of noise between different trainings
+* To simplify, why if you only keep the first tickers in terms of Market Cap?
+* Change the number of quarters to train in the past
+
+This function can be useful to compute the feature importance:
+
+
+```python
+def draw_feature_importance(model, top=15):
+    fi = model.feature_importance()
+    fn = model.feature_name()
+    feature_importance = pd.DataFrame(
+        [{"feature": fn[i], "imp": fi[i]} for i in range(len(fi))]
+    )
+    feature_importance = feature_importance.sort_values("imp", ascending=False).head(
+        top
+    )
+    feature_importance = feature_importance.sort_values("imp", ascending=True)
+    plot = (
+        ggplot(feature_importance, aes(x="feature", y="imp"))
+        + geom_col(fill="lightblue")
+        + coord_flip()
+        + scale_x_discrete(limits=feature_importance["feature"])
+    )
+    return plot
+```
+
+# Solution
+
+The solution propose will encompass the following structure:
+1. Define baseline: we will create a baseline model for performance comparison, which would be the 500 largest market cap tickers. We could even decide that    the SP500 itself might be the benchmark, but will use the same baseline as the proposed in the solution.
```

Comment:
> We could even decide that    the SP500 itself might be the benchmark

we could do this indeed, but this way we are creating our own "index" and we can adapt the code to equities which can't be compared against sp500